sklearn/cluster/k_medoids_.py

```diff
@@ -0,0 +1,284 @@
+
+    def fit(self, X, y=None):
+        """Fit K-Medoids to the provided data.
+
+        Parameters
+        ----------
+        X : {array-like, sparse matrix}, shape = (n_samples, n_features).
+            Dataset to cluster.
+
+        Returns
+        -------
+        self
+        """
+
+        self._check_init_args()
+        X = check_array(X, accept_sparse=['csr', 'csc'])
+        if self.n_clusters > X.shape[0]:
+            raise ValueError("The number of medoids %d must be less "
+                             "than the number of samples %d."
+                             % (self.n_clusters, X.shape[0]))
+
+        distances = pairwise_distances(X, metric=self.distance_metric)
+
+        medoid_idxs = self._get_initial_medoid_indices(distances,
+                                                       self.n_clusters)
+        labels = np.argmin(distances[medoid_idxs, :], axis=0)
+
+        # Old medoids will be stored here for reference
+        old_medoid_idxs = np.zeros((self.n_clusters,))
+
+        # Continue the algorithm as long as
+        # the medoids keep changing and the maximum number
+        # of iterations is not exceeded
+        self.n_iter_ = 0
+        while (not np.all(old_medoid_idxs == medoid_idxs) and
+                self.n_iter_ < self.max_iter):
+            self.n_iter_ += 1
+            old_medoid_idxs = np.copy(medoid_idxs)
+            labels = np.argmin(distances[medoid_idxs, :], axis=0)
+
+            # Update medoids with the new cluster indices
+            self._update_medoid_idxs_in_place(distances, labels, medoid_idxs)
+
+        # Expose labels_ which are the assignments of
+        # the training data to clusters
+        self.labels_ = labels
+        self.cluster_centers_ = X[medoid_idxs]
+        self.inertia_ = self._compute_inertia(X)
+
+        # Return self to enable method chaining
+        return self
+
+    def _update_medoid_idxs_in_place(self, distances, cluster_idxs,
+                                     medoid_idxs):
+        """In-place update of the medoid indices"""
+
+        # Update the medoids for each cluster
+        for k in range(self.n_clusters):
+
+            if sum(cluster_idxs == k) == 0:
+                warnings.warn("Cluster %d is empty!" % k)
+                continue
+
+            # Extract the distance matrix between the data points
+            # inside the cluster k
+            cluster_k_idxs = np.where(cluster_idxs == k)[0]
+            in_cluster_distances = distances[np.ix_(cluster_k_idxs,
+                                                    cluster_k_idxs)]
+
+            # Calculate all costs from each point to all others in the cluster
+            in_cluster_all_costs = np.sum(in_cluster_distances, axis=1)
+
+            min_cost_idx = np.argmin(in_cluster_all_costs)
+            min_cost = in_cluster_all_costs[min_cost_idx]
+            curr_cost = np.sum(distances[medoid_idxs[k], cluster_k_idxs])
```

Comment:
This value must be the same as a value from `in_cluster_all_costs`. Is it more sensible to pull it from there rather than recompute it? (`in_cluster_all_costs[np.argmax(cluster_k_idxs == medoid_idxs[k])]`)

Reply:
fixed